File: QuadTree.cpp
#include "QuadTree.h"



// constructor
QuadTree::QuadTree(Frontera frontera, int capacidad) {

    limite = frontera;
    capacidadMaxima = (size_t)capacidad;
    dividido = false;

    noroeste = nullptr;
    noreste = nullptr;
    suroeste = nullptr;
    sureste = nullptr;

}

// destructor
QuadTree::~QuadTree() {


    if (dividido) {

        delete noroeste;
        delete noreste;
        delete suroeste;
        delete sureste;
    }

}



// subdivide el nodo en 4 regiones
void QuadTree::subdividir() {


    // calculamos la dimension de los hijos
    double x = limite.x;
    double y = limite.y;
    double w = limite.ancho;
    double h = limite.alto;

    Frontera f_no = {x - w/2, y + h/2, w/2, h/2};
    Frontera f_ne = {x + w/2, y + h/2, w/2, h/2};
    Frontera f_so = {x - w/2, y - h/2, w/2, h/2};
    Frontera f_se = {x + w/2, y - h/2, w/2, h/2};


    // creamos los quadTrees
    noroeste = new QuadTree(f_no, (int)capacidadMaxima);
    noreste  = new QuadTree(f_ne, (int)capacidadMaxima);
    suroeste = new QuadTree(f_so, (int)capacidadMaxima);
    sureste  = new QuadTree(f_se, (int)capacidadMaxima);


    // prendemos la bandera
    dividido = true;

    // hacemos que las particulas del padre vengan a los nodos hijos segun su posicion
    for (const auto& p : particulas) {

        if (noroeste->insertar(p)) continue;
        if (noreste->insertar(p)) continue;
        if (suroeste->insertar(p)) continue;
        if (sureste->insertar(p)) continue;

    }

    particulas.clear();
}

// insertamos particulas, si se rebalsa, redistribuimos las partiuclas del nodo padre en los nodos hijos, por ultimo
// la particula que rebalsa, tambien lo ponemos en uno del los nodos hijos.
bool QuadTree::insertar(const Particle& p) {


    // no esta en el limite retornamos false
    if (!limite.contiene(p.x, p.y)) {

        return false;
    }

    // si el nodo no fue dividido y todavia tiene capacidad pusheamos, de lo contrario estamos en el limite y subdividimos.
    if (!dividido) {

        if (particulas.size() < capacidadMaxima) {

            particulas.push_back(p);
            return true;

        } else {
            subdividir();
        }
    }


    // si llegamos aqui entonces rebalso y tambien reubicamos la particula en el nodo hijo
    if (noroeste->insertar(p)) return true;
    if (noreste->insertar(p)) return true;
    if (suroeste->insertar(p)) return true;
    if (sureste->insertar(p)) return true;


    // si ocurre un inconcordancia matematica retornamos false, como si no se hubiera insertado
    return false;
}



// consultamos los nodos que estan en el rango rectangular, guardamos los encontrados y las comparaciones.
void QuadTree::consultarRango(const Frontera& rango, vector<Particle>& encontrados, int& comparacionesQuadTree) const {

    // si no esta en el rango , no buscamos
    if (!limite.intersecta(rango)) {
        return;
    }

    // si es hoja entonces buscamso ahi. todos las particulas estan en la hojas.
    if (!dividido) {

        for (const auto& p : particulas) {

            // anotamos las comparaciones
            comparacionesQuadTree++;

            // verificamos si se encuentra en el rango y pusheamos
            if (rango.contiene(p.x, p.y)) {

                encontrados.push_back(p);
            }
        }
    }
    // si llegamos aqui entonces no es hoja  y seguimos bajando por los nodos hijos y cuando llegemos a hoja se consulta
    else {


        noroeste->consultarRango(rango, encontrados, comparacionesQuadTree);
        noreste->consultarRango(rango, encontrados, comparacionesQuadTree);
        suroeste->consultarRango(rango, encontrados, comparacionesQuadTree);
        sureste->consultarRango(rango, encontrados, comparacionesQuadTree);
    }
}


// consultamos las particulas dentro de un circulo (px, py, radioBuscado).
// Filtra por distancia euclidiana, evitando los falsos positivos de las esquinas de una caja.
void QuadTree::consultarRadio(double px, double py, double radioBuscado, vector<Particle>& encontrados, int& comparacionesQuadTree) const {

    // creamos una caja que cubra el circulo buscado para el test de interseccion con el cuadrante
    Frontera rango = {px, py, radioBuscado, radioBuscado};

    // si la caja no intersecta con este cuadrante, lo descartamos
    if (!limite.intersecta(rango)) {

        return;
    }

    if (!dividido) {

        for (const auto& p : particulas) {

            comparacionesQuadTree++;

            // verificamos si la particula esta dentro del circulo
            double dx = p.x - px;
            double dy = p.y - py;
            double distanciaCuadrada = (dx * dx) + (dy * dy);

            if (distanciaCuadrada <= (radioBuscado * radioBuscado)) {

                encontrados.push_back(p);

            }
        }
    }
    else {

        noroeste->consultarRadio(px, py, radioBuscado, encontrados, comparacionesQuadTree);
        noreste->consultarRadio(px, py, radioBuscado, encontrados, comparacionesQuadTree);
        suroeste->consultarRadio(px, py, radioBuscado, encontrados, comparacionesQuadTree);
        sureste->consultarRadio(px, py, radioBuscado, encontrados, comparacionesQuadTree);
    }
}


// limpiamos todo el quadtree dejando el root reutilizable (sin reasignar memoria del root).
void QuadTree::limpiar() {


    particulas.clear();

    if (dividido) {

        noroeste->limpiar();
        noreste->limpiar();
        suroeste->limpiar();
        sureste->limpiar();

        delete noroeste; noroeste = nullptr;
        delete noreste; noreste = nullptr;
        delete suroeste; suroeste = nullptr;
        delete sureste; sureste = nullptr;

        dividido = false;
    }
}

File: web/js/QuadTree.js
// Espejo de QuadTree.cpp. Mismos nombres de metodo para validar paridad con el C++.
class QuadTree {
    constructor(frontera, capacidad) {
        this.limite = frontera;
        this.capacidadMaxima = capacidad;
        this.particulas = [];
        this.dividido = false;

        this.noroeste = null;
        this.noreste = null;
        this.suroeste = null;
        this.sureste = null;
    }

    subdividir() {
        const x = this.limite.x;
        const y = this.limite.y;
        const w = this.limite.ancho;
        const h = this.limite.alto;

        const f_no = new Frontera(x - w / 2, y + h / 2, w / 2, h / 2);
        const f_ne = new Frontera(x + w / 2, y + h / 2, w / 2, h / 2);
        const f_so = new Frontera(x - w / 2, y - h / 2, w / 2, h / 2);
        const f_se = new Frontera(x + w / 2, y - h / 2, w / 2, h / 2);

        this.noroeste = new QuadTree(f_no, this.capacidadMaxima);
        this.noreste  = new QuadTree(f_ne, this.capacidadMaxima);
        this.suroeste = new QuadTree(f_so, this.capacidadMaxima);
        this.sureste  = new QuadTree(f_se, this.capacidadMaxima);

        this.dividido = true;

        for (const p of this.particulas) {
            if (this.noroeste.insertar(p)) continue;
            if (this.noreste.insertar(p)) continue;
            if (this.suroeste.insertar(p)) continue;
            if (this.sureste.insertar(p)) continue;
        }

        this.particulas = [];
    }

    insertar(p) {
        if (!this.limite.contiene(p.x, p.y)) {
            return false;
        }

        if (!this.dividido) {
            if (this.particulas.length < this.capacidadMaxima) {
                this.particulas.push(p);
                return true;
            } else {
                this.subdividir();
            }
        }

        if (this.noroeste.insertar(p)) return true;
        if (this.noreste.insertar(p)) return true;
        if (this.suroeste.insertar(p)) return true;
        if (this.sureste.insertar(p)) return true;

        return false;
    }

    // query rectangular (primitivo canonico, util para culling de viewport)
    consultarRango(rango, encontrados, contador) {
        if (!this.limite.intersecta(rango)) {
            return;
        }

        if (!this.dividido) {
            for (const p of this.particulas) {
                contador.n++;
                if (rango.contiene(p.x, p.y)) {
                    encontrados.push(p);
                }
            }
        } else {
            this.noroeste.consultarRango(rango, encontrados, contador);
            this.noreste.consultarRango(rango, encontrados, contador);
            this.suroeste.consultarRango(rango, encontrados, contador);
            this.sureste.consultarRango(rango, encontrados, contador);
        }
    }

    // query circular: filtra por distancia euclidiana (correcto para colisiones por radio)
    consultarRadio(px, py, radioBuscado, encontrados, contador) {
        const rango = new Frontera(px, py, radioBuscado, radioBuscado);

        if (!this.limite.intersecta(rango)) {
            return;
        }

        if (!this.dividido) {
            for (const p of this.particulas) {
                contador.n++;
                const dx = p.x - px;
                const dy = p.y - py;
                const distanciaCuadrada = dx * dx + dy * dy;
                if (distanciaCuadrada <= radioBuscado * radioBuscado) {
                    encontrados.push(p);
                }
            }
        } else {
            this.noroeste.consultarRadio(px, py, radioBuscado, encontrados, contador);
            this.noreste.consultarRadio(px, py, radioBuscado, encontrados, contador);
            this.suroeste.consultarRadio(px, py, radioBuscado, encontrados, contador);
            this.sureste.consultarRadio(px, py, radioBuscado, encontrados, contador);
        }
    }

    limpiar() {
        this.particulas = [];
        if (this.dividido) {
            this.noroeste.limpiar();
            this.noreste.limpiar();
            this.suroeste.limpiar();
            this.sureste.limpiar();
            this.noroeste = null;
            this.noreste = null;
            this.suroeste = null;
            this.sureste = null;
            this.dividido = false;
        }
    }

    // dibuja las lineas de subdivision; colorea hojas por densidad (#11, #19)
    dibujar(ctx, capacidad) {
        const x0 = this.limite.x - this.limite.ancho;
        const y0 = this.limite.y - this.limite.alto;
        const w = this.limite.ancho * 2;
        const h = this.limite.alto * 2;

        if (!this.dividido) {
            // relleno por densidad: mas particulas -> mas opaco
            const dens = Math.min(1, this.particulas.length / capacidad);
            if (dens > 0) {
                ctx.fillStyle = `rgba(0, 220, 255, ${0.05 + dens * 0.25})`;
                ctx.fillRect(x0, y0, w, h);
            }
        }

        ctx.strokeStyle = 'rgba(0, 220, 255, 0.25)';
        ctx.lineWidth = 0.5;
        ctx.strokeRect(x0, y0, w, h);

        if (this.dividido) {
            this.noroeste.dibujar(ctx, capacidad);
            this.noreste.dibujar(ctx, capacidad);
            this.suroeste.dibujar(ctx, capacidad);
            this.sureste.dibujar(ctx, capacidad);
        }
    }
}
